### PR TITLE
[OD-1177] Add FGDC harvester

### PR DESCRIPTION
**Pull request by jguo144**
## Description
This PR adds a harvester to get metadata from a CSW endpoint using the FGDC schema.

## Testing
- Install the extension and restart supervisor
```
git checkout jguo144/OD-1177/2019-03-15/fgdc-parser
python setup.py develop
pip install -r pip-requirements.txt
sudo service supervisor restart
```
- Harvest the link below using `CSW FGDC Server` harvester
http://pycswtest.cnra.ca.gov/servcat_fws/csw.py?&config=servcat_fws_profile.cfg&service=CSW&version=2.0.2&request=GetCapabilities

### CODE REVIEW

File: pip-requirements.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 GeoAlchemy>=0.6
-GeoAlchemy2>=0.2.4
+GeoAlchemy2==0.5.0
```
Comment (ankitbrana) [inline pip-requirements.txt:2]: I'm assuming this change is required to support the new harvester?

File: ckanext/spatial/harvesters/csw_fgdc.py [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+        return url
+
+    def output_schema(self):
+        return 'fgdc'
+
+    def gather_stage(self, harvest_job):
+        log = logging.getLogger(__name__ + '.CSW.gather')
+        log.debug('CswHarvester gather_stage for job: %r', harvest_job)
+        # Get source URL
+        url = harvest_job.source.url
```
Comment (ankitbrana) [inline ckanext/spatial/harvesters/csw_fgdc.py:135]: Do we need a null/undefined check for source?

File: ckanext/spatial/harvesters/csw_fgdc.py [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+            p.toolkit.get_action('package_delete')(context, {'id': harvest_object.package_id})
+            log.info('Deleted package {0} with guid {1}'.format(harvest_object.package_id, harvest_object.guid))
+
+            return True
+
+        # Parse as FGDC document
+        try:
+            fgdc_parser = FGDCDocument(harvest_object.content)
+            fgdc_values = fgdc_parser.read_values()
+            log.debug(fgdc_values)
```
Comment (ankitbrana) [inline ckanext/spatial/harvesters/csw_fgdc.py]: What is `fgdc_values`, how big is the log?

File: ckanext/spatial/harvesters/csw_fgdc.py [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+            m = hashlib.md5()
+            m.update(harvest_object.content.encode('utf8', 'ignore'))
+            harvest_object.guid = m.hexdigest()
+            harvest_object.add()
+
+        # Get document modified date
+        try:
+            metadata_modified_date = dateutil.parser.parse(fgdc_values['metadata-date'], ignoretz=True)
+        except ValueError:
+            self._save_object_error('Could not extract reference date for object {0} ({1})'
```
Comment (ankitbrana) [inline ckanext/spatial/harvesters/csw_fgdc.py:337]: Just to clarify, is this error being logged in the harvest logs?

File: ckanext/spatial/harvesters/csw_fgdc.py [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+        # Build the package dict
+        package_dict = self.get_package_dict(fgdc_values, harvest_object)
+        for harvester in p.PluginImplementations(ISpatialHarvester):
+            package_dict = harvester.get_package_dict(context, {
+                'package_dict': package_dict,
+                'fgdc_values': fgdc_values,
+                'xml_tree': fgdc_parser.xml_tree,
+                'harvest_object': harvest_object,
+            })
+        if not package_dict:
```
Comment (ankitbrana) [inline ckanext/spatial/harvesters/csw_fgdc.py:354]: Is this if statement properly aligned, should it be within the for loop?

File: ckanext/spatial/harvesters/csw_fgdc.py [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+           'extras_as_string': True,
+           'api_version': '2',
+           'return_id_only': True})
+
+        if self._site_user and context['user'] == self._site_user['name']:
+            context['ignore_auth'] = True
+
+
+        # The default package schema does not like Upper case tags
+        tag_schema = logic.schema.default_tags_schema()
```
Comment (ankitbrana) [inline ckanext/spatial/harvesters/csw_fgdc.py:369]: Do we need a checck for schema being undefined/null?

**Review by ankitbrana — COMMENTED** [review]
Couple questions, non blocking

File: pip-requirements.txt (hunk 1 above)
Comment (ankitbrana) [reply pip-requirements.txt:2]: Resolved

File: ckanext/spatial/harvesters/csw_fgdc.py (hunk 3 above)
Comment (jguo144) [reply ckanext/spatial/harvesters/csw_fgdc.py]: Good catch. I'll remove the log message.

**Pull request merged**